### PR TITLE
Remove long type in DQMServices and Integration [12_1_X]

### PR DESCRIPTION
**Pull request by jfernan2**
#### PR description:

Forward port of https://github.com/cms-sw/cmssw/pull/35307

### CODE REVIEW

**Comment by jfernan2** [comment]
please test

File: DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py [hunk 1]
```diff
@@ -243,7 +243,7 @@
     process.SiStripAnalyserCosmic.TkMapCreationFrequency  = -1
     process.SiStripAnalyserCosmic.ShiftReportFrequency = -1
     process.SiStripAnalyserCosmic.StaticUpdateFrequency = 5
-    process.SiStripAnalyserCosmic.MonitorSiStripBackPlaneCorrection = False
+    process.SiStripAnalyserCosmic.MonitorSiStripBackPlaneCorrection = cms.bool(False)
```
Comment (mmusich) [inline DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py:246]: why is this needed?

Comment (jfernan2) [reply DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py:246]: It looks like it is indeed needed because that parameter is not set in SiStripAnalyserCosmic module but inherited from another config.
We (@pmandrik and I) found these errors when running DQM clients in python3 on 12_0_X for the very first time. It seems that dropping type specs only works if the parameters are explicitely set in the cloned modules, those inherited via C++ class inheritance or deference (like this) are not taken.
However, I am not a python expert, so any advice is welcomed, specially because this looks like a general problem, I wonder if nobody has faced it before us.

Comment (mmusich) [reply DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py:246]: @makortel

Comment (makortel) [reply DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py:246]: I see the `process.SiStripAnalyserCosmic` apparently comes from
https://github.com/cms-sw/cmssw/blob/a19e4f667b438dc92d99069a3dfd5de526b52db3/DQM/SiStripMonitorClient/python/SiStripClientConfigP5_Cosmic_cff.py#L5
and that indeed does not have `MonitorSiStripBackPlaneCorrection` parameter defined. Therefore setting that paremeter here requires the explicit type. Perhaps the parameter should be defined in `DQM.SiStripMonitorClient.SiStripClientConfigP5_Cosmic_cff`? Then setting it here would not need an explicit type. (the cfi fragments should really contain the complete set of parameters).

I'm not really sure what exactly you meant with
> parameter ... inherited from another config

or
> those inherited via C++ class inheritance or deference (like this) are not taken.

-------

(as a side note, the `SiStripAnalyzer` appears to be a legacy module and needs to be migrated to one of the thread-friendly module types)

Comment (mmusich) [reply DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py:246]: > perhaps the parameter should be defined in DQM.SiStripMonitorClient.SiStripClientConfigP5_Cosmic_cff?

indeed, but I was wondering why it worked in py2. Or did it? @jfernan2 

> (as a side note, the SiStripAnalyzer appears to be a legacy module and needs to be migrated to one of the thread-friendly module types)

That's on my todo list.

Comment (jfernan2) [reply DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py:246]: > indeed, but I was wondering why it worked in py2. Or did it? @jfernan2
It did not work, well, we did not even try since these type spec removal was done at 12_0_X when it was master and we are actually testing it now at P5

Comment (jfernan2) [reply DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py:246]: > I'm not really sure what exactly you meant with

I meant cases like this or this other example modified in this PR:
DQM/EcalCommon/python/CommonParams_cfi.py
DQM/Integration/python/clients/hcalcalib_dqm_sourceclient-live_cfg.py

where the parameter subsystem is defined indirectly though C++ class inheritance, and set in the corresponding config of the original mother class, but not in the son/grandson cfi

Comment (makortel) [reply DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py:246]: Could you be more specific? The `DQM/EcalCommon/python/CommonParams_cfi.py` does not have anything that would relate to C++ class inheritance, and `DQM/Integration/python/clients/hcalcalib_dqm_sourceclient-live_cfg.py` has too many things to really get a grasp.

Regardless of the structure of the C++ code, it is the responsibility of the concrete EDModule to define its full cfi fragment, either via its `fillDescriptions()` function (preferred), or in a hand-crafted `cfi.py` file.

Comment (jfernan2) [reply DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py:246]: Sure, I will focus on:
DQM/Integration/python/clients/hcalcalib_dqm_sourceclient-live_cfg.py
L186 and L195 modify subsytem parameter in module qie11Task_laser = process.qie11Task.clone()

qie11Task is defined in:
https://github.com/cms-sw/cmssw/blob/edb013d63e3b1a934c974d310587366e3f5b242c/DQM/HcalTasks/python/QIE11Task.py

which is an instance of C++ plugin QIE11Task 
https://github.com/cms-sw/cmssw/blob/edb013d63e3b1a934c974d310587366e3f5b242c/DQM/HcalTasks/plugins/QIE11Task.cc

but does not set subsystem parameter because this parameter comes from inheritance of DQTask class:
https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/DQM/HcalCommon/src/DQTask.cc

which does NOT set subsystem parameter either because it inherits too from class DQModule:
https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/DQM/HcalCommon/src/DQModule.cc#L11
where there is a default value but at C++ code.

So, I could go and create a cfi for DQModule and set default values which are at C++ code, and then start calling it from python config files here and there...but I do not know if this is recommended at all...

Any advice is welcomed...

Comment (makortel) [reply DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py:246]: Thanks for the explanation.

In the end the parameters need to end up in `DQM/HcalTasks/python/QIE11Task.py`. This can be achieved in (at least) three ways:

1. Add the parameters used by `QIE11Task`, `DQTask`, and `DQModule` manually to `qie11Task`. Leads to lot's of repetition if there are many other classes inheriting from `DQTask` and `DQModule`.
2. Create (handcrafted) PSets for `DQTask` and `DQModule`, and propagate the parameters from there to `qie11Task`. Eliminates the repetition of parameters, but exposes the inheritance structure in the configuration, that likely wouldn't need to know it.
3. Add `fillDescriptions()` function to `QIE11Task`, and `fillPSetDescription()` functions to `DQTask` and `DQModule`, such that the deriving class' function calls the parent class' function to add the parent class' parameters. Eliminates repetition, and the code organization (class hierarchy) stays as an implementation detail of `QIE11Task`.

The last one (3) would be the preferred pattern.

Comment (jfernan2) [reply DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py:246]: Thanks @makortel 
Could you point me to a good example?
I have found this twiki:
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#Converting_an_Existing_CFI_file
and I am able to create fillDescriptions() function to QIE11Task but I am not sure if it should go to .h file (protected, public section) or to .cc
And in case of the parent classes, I am not sure how to create fillPSetDescription() functions.
Thanks

Comment (jfernan2) [reply DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py:246]: I have pushed my naive attempt to fulfill the comments, any advice is welcomed!
Thanks

**Comment by jfernan2** [comment]
please test

File: DQM/HcalTasks/plugins/QIE11Task.cc [hunk 2]
```diff
@@ -18,6 +18,27 @@ QIE11Task::QIE11Task(edm::ParameterSet const& ps)
...
+  desc.addUntracked<int>("runkeyVal", 0);
+  desc.addUntracked<std::string>("runkeyName", "pp_run");
+  desc.addUntracked<edm::InputTag>("tagQIE11", edm::InputTag("hcalDigis"));
+  desc.addUntracked<double>("cut", 20);
+  desc.addUntracked<int>("ped", 4);
+  desc.addUntracked<int>("laserType", -1);
+  desc.addUntracked<int>("eventType", -1);
+  descriptions.add("qie11Task", desc);
+  // or use the following to generate the label from the module's C++ type
+  //descriptions.addWithDefaultLabel(desc);
```
Comment (makortel) [inline DQM/HcalTasks/plugins/QIE11Task.cc:39]: I'd suggest to remove these lines if you're not going to use the `addWithDefaultLabel()`
```suggestion
```

**Comment by jfernan2** [comment]
Thanks a lot as usual @makortel !

> Eventually in the configuration side the `DQM.HcalTasks.qie11Task_cfi` should be used instead of the current `DQM.HcalTasks.QIE11Task_cfi` (or the latter can import the former if migrating all users would be deemed too laborious).

For this I decided to adopt another strategy using addDefault since there are several clones of the same cfi, I do not know what you think.

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
please test

File: DQM/SiStripMonitorClient/python/SiStripClientConfigP5_cff.py [hunk 3]
```diff
@@ -36,7 +36,7 @@
     ),
 
 # Parameters for On Demand Monitoring   
-    MonitorSiStripBackPlaneCorrection = OnDemandMonitoring.MonitorSiStripBackPlaneCorrection,
+    MonitorSiStripBackPlaneCorrection = cms.bool(OnDemandMonitoring.MonitorSiStripBackPlaneCorrection),
```
Comment (jfernan2) [inline DQM/SiStripMonitorClient/python/SiStripClientConfigP5_cff.py:39]: @makortel  I need help with this one too: this does not work and crashes with 
ValueError: cms.bool(False) is not a valid <class 'FWCore.ParameterSet.Types.bool'>  (???)
I don't understand why, I tried using eval but it does not work either.

Comment (makortel) [reply DQM/SiStripMonitorClient/python/SiStripClientConfigP5_cff.py:39]: The problem is that `OnDemandMonitoring.MonitorSiStripBackPlaneCorrection` is a  `cms.bool()` itself, and `cms.bool(cms.bool(False))` indeed does not work. The following would work
```suggestion
    MonitorSiStripBackPlaneCorrection = cms.bool(OnDemandMonitoring.MonitorSiStripBackPlaneCorrection.value()),
```

Note that below lines like
```py
    MonitorSiStripPedestal      = OnDemandMonitoring.MonitorSiStripPedestal,
```
make `SiStripAnalyser.MonitorSiStripPedestal` and `OnDemandMonitoring.MonitorSiStripPedestal` be the same object, so if someone changes `SiStripAnalyser.MonitorSiStripPedestal` the `OnDemandMonitoring.MonitorSiStripPedestal` changes as well (since they are the same). We typically discourage these kind of "long distance interactions" because they can cause very confusing situations to debug.

One way to reduce boilerplate would be to define a PSet for the options whose names and values are the same between `SiStripAnalyser` and `OnDemandMonitoring`: move those to a PSet that is then used to initialize both EDModule configurations
```
common = cms.PSet(...)
SiStripAnalyser = cms.EDAnalyzer("SiStripAnalyser", common, ...)
OnDemandMonitoring = cms.EDAnalyzer("SiStripMonitorCondDataOnDemandExample", common, ...)
```
The parameters from the `common` PSet are copied to both EDAnalyzers.

Comment (jfernan2) [reply DQM/SiStripMonitorClient/python/SiStripClientConfigP5_cff.py]: Thanks!
About the common PSet, since SiStripAnalyser and OnDemandMonitoring are defined in different cfi files, I am sorry for my naive question but, I am not sure how to handle it then

**Comment by jfernan2** [comment]
please test

File: DQM/SiStripMonitorClient/python/SiStripClientConfigP5_cff.py (hunk 3 above)
Comment (makortel) [reply DQM/SiStripMonitorClient/python/SiStripClientConfigP5_cff.py]: You can define the "common PSet" (btw, please use something more descriptive than my `common`) e.g. in a new cfi file that gets imported in `SiStripClientConfigP5_cff` and `OnDemandMonitoring_cfi', or just define it in `OnDemandMonitoring_cfi' since that one is already imported into `SiStripClientConfigP5_cff`.

Comment (jfernan2) [reply DQM/SiStripMonitorClient/python/SiStripClientConfigP5_cff.py]: Thanks @makortel 
I tried to implement your suggestion but I have the impression that, unless mistake from my side, the code gets more complicated due to the number of changes involved....

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
please test
Not able to reproduce the error in my local setup...repeating test

**Comment by jfernan2** [comment]
+1

File: DQM/SiStripMonitorSummary/src/SiStripClassToMonitorCondData.cc [hunk 4]
```diff
@@ -51,15 +51,17 @@
...
-  monitorPedestals_ = iConfig.getParameter<bool>("MonitorSiStripPedestal");
-  monitorNoises_ = iConfig.getParameter<bool>("MonitorSiStripNoise");
-  monitorQuality_ = iConfig.getParameter<bool>("MonitorSiStripQuality");
-  monitorApvGains_ = iConfig.getParameter<bool>("MonitorSiStripApvGain");
-  monitorLorentzAngle_ = iConfig.getParameter<bool>("MonitorSiStripLorentzAngle");
-  monitorBackPlaneCorrection_ = iConfig.getParameter<bool>("MonitorSiStripBackPlaneCorrection");
-  monitorLowThreshold_ = iConfig.getParameter<bool>("MonitorSiStripLowThreshold");
-  monitorHighThreshold_ = iConfig.getParameter<bool>("MonitorSiStripHighThreshold");
-  monitorCabling_ = iConfig.getParameter<bool>("MonitorSiStripCabling");
+  const auto& mPSet = iConfig.getParameter<edm::ParameterSet>("MonitorSiStrip_PSet");
```
Comment (makortel) [inline DQM/SiStripMonitorSummary/src/SiStripClassToMonitorCondData.cc:54]: Sorry to stir the pot more, but I just want to point out that if you would do
```py
SiStripAnalyserHI = cms.EDAnalyzer("SiStripAnalyser",
    OnDemandMonitoring.MonitorSiStrip_PSet,
    StaticUpdateFrequency    = cms.untracked.int32(1),
```
instead of current
```py
SiStripAnalyserHI = cms.EDAnalyzer("SiStripAnalyser",
    StaticUpdateFrequency    = cms.untracked.int32(1),
    ....
    MonitorSiStrip_PSet          = OnDemandMonitoring.MonitorSiStrip_PSet.clone(...)
```
these changes in the configuration reading side would not be needed.

On the other hand, there is already another PSet, `FillConditions_PSet`, so the change is not dramatically different from the existing content (and there may be other motivations for it).

Comment (jfernan2) [reply DQM/SiStripMonitorSummary/src/SiStripClassToMonitorCondData.cc:54]: Thanks @makortel 
I decided the current solution for the sake of not mixing the ordering of the parameters, which are many. So unless you have a strong opinion I would leave as it is

File: DQM/EcalCommon/python/CommonParams_cfi.py [hunk 5]
```diff
@@ -4,8 +4,8 @@
     onlineMode = cms.untracked.bool(False),
     willConvertToEDM = cms.untracked.bool(True)
 )
-ecaldqmLaserWavelengths = 1, 2, 3
-ecaldqmLedWavelengths = 1, 2
-ecaldqmMGPAGains = 12
-ecaldqmMGPAGainsPN = 16
 
+ecaldqmLaserWavelengths = cms.untracked.vint32(1, 2, 3)
```
Comment (perrotta) [inline DQM/EcalCommon/python/CommonParams_cfi.py:8]: A safer way for updating those vectors of int is:
```suggestion
ecaldqmLaserWavelengths = [1, 2, 3]
ecaldqmLedWavelengths = [1, 2]
ecaldqmMGPAGains = [12]
ecaldqmMGPAGainsPN = [16]
```

Comment (jfernan2) [reply DQM/EcalCommon/python/CommonParams_cfi.py]: Thanks

Comment (jfernan2) [reply DQM/EcalCommon/python/CommonParams_cfi.py]: Sorry, I realized now your suggestion is a bit weird, we need to set them as cms.untracked here so, for my understanding, what do the brackets do in the previous lines?

Comment (perrotta) [reply DQM/EcalCommon/python/CommonParams_cfi.py]: Sorry, my fault!
I had the impression that you were updating already existing parameters, not defining some new ones in the PSet (I was looking a the updated code too quickly, my bad...).
Please revert the last commit, so that we can merge this as it was

File: DQM/SiStripMonitorSummary/src/SiStripClassToMonitorCondData.cc (hunk 4 above)
Comment (makortel) [reply DQM/SiStripMonitorSummary/src/SiStripClassToMonitorCondData.cc:54]: No strong feelings, just wanted to tell what is possible to do.

File: DQM/EcalCommon/python/CommonParams_cfi.py (hunk 5 above)
Comment (jfernan2) [reply DQM/EcalCommon/python/CommonParams_cfi.py:8]: Done. Thanks!

**Comment by perrotta** [comment]
+1
- Only one commit was added and then reverted since when this PR was lastly tested, therefore we can avoid relaunching the tests here

**Comment by perrotta** [comment]
merge

**Comment by perrotta** [comment]
please abort test
